src/HttpClientFactory/Http/test/DependencyInjection/HttpClientFactoryServiceCollectionExtensionsTest.cs

```diff
@@ -996,5 +1020,149 @@ public TypedClientWithTransientService(HttpClient httpClient, TransientService s
 
             public TransientService Service { get; }
         }
+
+        // https://github.com/nunit/nunit/blob/master/src/NUnitFramework/framework/Internal/SingleThreadedSynchronizationContext.cs
+        private class SingleThreadedSynchronizationContext : SynchronizationContext, IDisposable
```

Comment:
Just use blocking collection, this is a bit overengineered for a unit test.

Reply:
because the simple one `SingleThreadedSynchronizationContext.cs` cannot re-produce same deadlocks behavior.

```csharp
using System;
using System.Collections.Concurrent;
using System.Net.Http;
using System.Threading;
using Microsoft.Extensions.DependencyInjection;

namespace Console1
{
    public class Program
    {
        static void Main(string[] args)
        {
            var services = new ServiceCollection();
            services.AddHttpClient();
            var provider = services.BuildServiceProvider();
            var factory = provider.GetRequiredService<IHttpClientFactory>();
            var client1 = new HttpClient();
            SingleThreadedSynchronizationContext.Run(() =>
            {
                Console.Write("default HttpClient in single-threaded SynchronizationContext ... ");
                client1.GetAsync("http://example.com").GetAwaiter().GetResult();
                Console.WriteLine($"Done");
            });
            var client2 = factory.CreateClient();
            SingleThreadedSynchronizationContext.Run(() =>
            {
                Console.Write("managed HttpClient in single-threaded SynchronizationContext ... ");
                client2.GetAsync("http://example.com").GetAwaiter().GetResult();
                Console.WriteLine($"Done");
            });
        }
    }

    public class SingleThreadedSynchronizationContext : SynchronizationContext
    {
        private readonly BlockingCollection<(SendOrPostCallback Callback, object State)> _queue = new BlockingCollection<(SendOrPostCallback Callback, object State)>();
        private readonly Thread _thread;

        public SingleThreadedSynchronizationContext()
        {
            _thread = new Thread(ProcessQueue);
            _thread.Start();
        }

        private void ProcessQueue(object obj)
        {
            foreach (var item in _queue.GetConsumingEnumerable())
            {
                item.Callback(item.State);
            }
        }

        public override void Post(SendOrPostCallback d, object state)
        {
            _queue.Add((d, state));
        }

        public static void Run(Action action)
        {
            var previous = SynchronizationContext.Current;
            var context = new SingleThreadedSynchronizationContext();
            SynchronizationContext.SetSynchronizationContext(context);
            try
            {
                action();
            }
            finally
            {
                SynchronizationContext.SetSynchronizationContext(previous);
            }
        }
    }
}
```

Run this code

```sh
default HttpClient in single-threaded SynchronizationContext ... Done
managed HttpClient in single-threaded SynchronizationContext ... Done
```

It will hangs on Console exiting.

Reply:
Thats because the thread needs to be a background thread. I'll test it out. It should be easy to reproduce

Reply:
Odd, when I run it, it never hits the sync context.

Reply:
Ah, wait my bad that was the first one.